### DISCUSSION TITLE
How to check for metavariable equality between mutiple patterns?

### DISCUSSION THREAD

**Discussion by voidpumpkin** [ast-grep/ast-grep#431] [Q&A]
I am trying to write a automatic linting rule for this:
```
# Avoid duplicated exports


Generally, we don't encourage the use of re-exports.

However, sometimes, to keep the interface exposed by a lib crate tidy, we use re-exports to shorten the path to specific items.
When doing so, a pitfall is to export a single item under two different names.

Consider:
    pub mod foo;
    pub use foo::Foo;

The issue with this code, is that `Foo` is now exposed under two different paths: `Foo`, `foo::Foo`.

This unnecessarily increases the surface of your API.
It can also cause issues on the client side. For example, it makes the usage of auto-complete in the IDE more involved.

Instead, ensure you export only once with `pub`.
```

My plan was to find `foo` in `pub use foo::Foo;` and check if  `pub mod foo;` exists in the same file.

Playground: https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InJ1c3QiLCJxdWVyeSI6IiIsImNvbmZpZyI6InJ1bGU6XG4gIGFsbDpcbiAgICAgLSBwYXR0ZXJuOiBwdWIgdXNlICQkJEE6OiRDO1xuICAgICAtIGluc2lkZTpcbiAgICAgICAgYWxsOlxuICAgICAgICAgIC0ga2luZDogc291cmNlX2ZpbGVcbiAgICAgICAgICAtIGhhczpcbiAgICAgICAgICAgICAgcGF0dGVybjogcHViIG1vZCAkJCRBOyIsInNvdXJjZSI6InB1YiBtb2QgZm9vO1xucHViIHVzZSBmb286OkZvbztcblxuXG5cbnB1YiBtb2QgZm9vO1xuXG5wdWIgdXNlIGFhYTo6QTtcblxuXG5cbnB1YiB1c2Ugd29vOjpXb287In0=

But I see that `$$$A` in both patterns are matched separately.

How to go about this then?

**Comment by HerringtonDarkholme** [comment]
Hi! Thanks for your continuous interest in ast-grep! I'm glad to see advanced usage of the tool! 

---

Multi Meta Variable currently is not checked currently, sadly. :( 

The main reason is that it is not clear to compare two sequences of nodes. For single meta variables like `$A`, I just check if the two AST nodes match exactly. But for a multiple meta var like `$$$A`, their contents are a list of nodes. The node list also includes unnamed nodes which is not important for matching purposes.

-----

That said, for the specific lint rule, we can use safely use `mod $A` since Rust only allow mod name to be `identifier`. See https://doc.rust-lang.org/reference/items/modules.html
![image](https://github.com/ast-grep/ast-grep/assets/2883231/4a587e9d-2b2e-4a9d-ab62-a3dcb601142a)

If we target this syntax, we can combine single meta var, `has` and `stopBy` operators.

```yaml
rule:
  all:
     - pattern: pub use $B::$C;
     - inside:
        kind: source_file
        has:
          pattern: pub mod $A;
     - has:
        pattern: $A
        stopBy: end
```

<img width="927" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/e2d2cdb5-389a-4d65-88cb-57a03ce62d63">

https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InJ1c3QiLCJxdWVyeSI6IiIsImNvbmZpZyI6InJ1bGU6XG4gIGFsbDpcbiAgICAgLSBwYXR0ZXJuOiBwdWIgdXNlICRCOjokQztcbiAgICAgLSBpbnNpZGU6XG4gICAgICAgIGtpbmQ6IHNvdXJjZV9maWxlXG4gICAgICAgIGhhczpcbiAgICAgICAgICBwYXR0ZXJuOiBwdWIgbW9kICRBO1xuICAgICAtIGhhczpcbiAgICAgICAgcGF0dGVybjogJEFcbiAgICAgICAgc3RvcEJ5OiBlbmQiLCJzb3VyY2UiOiJwdWIgbW9kIGZvbztcbnB1YiB1c2UgZm9vOjpGb287XG5wdWIgdXNlIGZvbzo6QTo6QjtcblxuXG5wdWIgdXNlIGFhYTo6QTtcbnB1YiB1c2Ugd29vOjpXb287In0=

**Reply by HerringtonDarkholme** [reply]
A stricter rule is to use the recursive rule `utils` and operator `matches`.  It is even more advanced but also more precise. https://ast-grep.github.io/guide/rule-config/utility-rule.html

https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InJ1c3QiLCJxdWVyeSI6IiIsImNvbmZpZyI6InJ1bGU6XG4gIGFsbDpcbiAgICAgLSBwYXR0ZXJuOiBwdWIgdXNlICRCOjokQztcbiAgICAgLSBpbnNpZGU6XG4gICAgICAgIGtpbmQ6IHNvdXJjZV9maWxlXG4gICAgICAgIGhhczpcbiAgICAgICAgICBwYXR0ZXJuOiBwdWIgbW9kICRBO1xuICAgICAtIGhhczpcbiAgICAgICAgbWF0Y2hlczogZmlyc3RcbnV0aWxzOlxuICBmaXJzdDpcbiAgICBraW5kOiBzY29wZWRfaWRlbnRpZmllclxuICAgIGhhczpcbiAgICAgIGZpZWxkOiBwYXRoXG4gICAgICBhbnk6XG4gICAgICAgIC0gcGF0dGVybjogJEFcbiAgICAgICAgLSBtYXRjaGVzOiBmaXJzdFxuIiwic291cmNlIjoicHViIG1vZCBmb287XG5wdWIgdXNlIGZvbzo6Rm9vO1xucHViIHVzZSBmb286OkE6OkI7XG5wdWIgdXNlIGZvbzo6YmFyOjpidXo6OmQ7XG5cbnB1YiB1c2Ugd29vOjpmb286OmE7XG5wdWIgdXNlIGFhYTo6QTtcbnB1YiB1c2Ugd29vOjpXb287In0=
